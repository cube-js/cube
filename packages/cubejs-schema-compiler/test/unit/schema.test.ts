import { prepareCompiler } from './PrepareCompiler';

describe('Schema Testing', () => {
  const schemaCompile = async () => {
    const { compiler, cubeEvaluator } = prepareCompiler(` 
      cube('CubeA', {
        sql: \`select * from test\`,
   
        measures: {
          count: {
            type: 'count'
          }
        },
  
        dimensions: {
          id: {
            type: 'number',
            sql: 'id',
            primaryKey: true
          },
          createdAt: {
            type: 'time',
            sql: 'created_at'
          },
        },

        preAggregations: {
            // Pre-aggregation without type, rollup is used by default
            countCreatedAt: {
                measureReferences: [count],
                timeDimensionReference: createdAt,
                granularity: \`day\`,
                partitionGranularity: \`month\`
            }
        },
      }) 
    `);
    await compiler.compile();

    return { compiler, cubeEvaluator };
  };

  it('valid schemas', async () => {
    const { cubeEvaluator } = await schemaCompile();

    expect(cubeEvaluator.preAggregationsForCube('CubeA')).toEqual({
      countCreatedAt: {
        external: false,
        scheduledRefresh: false,
        granularity: 'day',
        measureReferences: expect.any(Function),
        timeDimensionReference: expect.any(Function),
        partitionGranularity: 'month',
        type: 'rollup',
      }
    });
  });

  it('valid schemas (preview flags)', async () => {
    process.env.CUBEJS_EXTERNAL_DEFAULT = 'true';
    process.env.CUBEJS_SCHEDULED_REFRESH_DEFAULT = 'true';

    const { cubeEvaluator } = await schemaCompile();

    delete process.env.CUBEJS_EXTERNAL_DEFAULT;
    delete process.env.CUBEJS_SCHEDULED_REFRESH_DEFAULT;

    expect(cubeEvaluator.preAggregationsForCube('CubeA')).toEqual({
      countCreatedAt: {
        // because preview
        external: true,
        scheduledRefresh: true,
        granularity: 'day',
        measureReferences: expect.any(Function),
        timeDimensionReference: expect.any(Function),
        partitionGranularity: 'month',
        type: 'rollup',
      }
    });
  });
});
