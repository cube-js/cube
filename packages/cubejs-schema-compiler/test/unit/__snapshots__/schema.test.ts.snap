// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 1`] = `
Object {
  "completed_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "created_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "id": Object {
    "ownedByCube": true,
    "primaryKey": true,
    "sql": [Function],
    "type": "number",
  },
  "status": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "user_id": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "number",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 2`] = `
Object {
  "city": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "completed_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "created_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "id": Object {
    "ownedByCube": true,
    "primaryKey": true,
    "sql": [Function],
    "type": "number",
  },
  "status": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "user_id": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "number",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 3`] = `
Object {
  "count": Object {
    "ownedByCube": true,
    "type": "count",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 4`] = `
Object {
  "count": Object {
    "ownedByCube": true,
    "type": "count",
  },
  "count_distinct": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "countDistinct",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 5`] = `
Object {
  "sfUsers": Object {
    "description": "SF users segment from createCubeSchema",
    "ownedByCube": true,
    "sql": [Function],
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 6`] = `
Object {
  "anotherStatus": Object {
    "description": "Just another one",
    "ownedByCube": true,
    "sql": [Function],
  },
  "sfUsers": Object {
    "description": "SF users segment from createCubeSchema",
    "ownedByCube": true,
    "sql": [Function],
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 7`] = `
Object {
  "hello": Object {
    "levels": [Function],
    "title": "World",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 8`] = `
Object {
  "ehlo": Object {
    "levels": [Function],
    "title": "UnderGround",
  },
  "hello": Object {
    "levels": [Function],
    "title": "World",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 9`] = `
Object {
  "countCreatedAt": Object {
    "external": true,
    "granularity": "day",
    "measureReferences": [Function],
    "partitionGranularity": "month",
    "refreshKey": Object {
      "every": "1 hour",
    },
    "scheduledRefresh": true,
    "timeDimensionReference": [Function],
    "type": "rollup",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 10`] = `
Object {
  "countCreatedAt": Object {
    "external": true,
    "granularity": "day",
    "measureReferences": [Function],
    "partitionGranularity": "month",
    "refreshKey": Object {
      "every": "1 hour",
    },
    "scheduledRefresh": true,
    "timeDimensionReference": [Function],
    "type": "rollup",
  },
  "mainPreAggs": Object {
    "dimensionReferences": [Function],
    "external": true,
    "measureReferences": [Function],
    "scheduledRefresh": true,
    "type": "rollup",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 11`] = `
Array [
  Object {
    "role": "*",
    "rowLevel": Object {
      "allowAll": true,
    },
  },
  Object {
    "conditions": Array [
      Object {
        "if": [Function],
      },
    ],
    "memberLevel": Object {
      "excludes": Array [
        "status",
      ],
      "excludesMembers": Array [
        "ordersExt.status",
      ],
      "includes": "*",
      "includesMembers": Array [
        "ordersExt.count",
        "ordersExt.count_distinct",
        "ordersExt.id",
        "ordersExt.user_id",
        "ordersExt.status",
        "ordersExt.created_at",
        "ordersExt.completed_at",
        "ordersExt.city",
        "ordersExt.sfUsers",
        "ordersExt.anotherStatus",
      ],
    },
    "role": "admin",
    "rowLevel": Object {
      "filters": Array [
        Object {
          "member": [Function],
          "memberReference": "ordersExt.id",
          "operator": "equals",
          "values": [Function],
        },
      ],
    },
  },
]
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.js (with additions) 12`] = `
Array [
  Object {
    "role": "*",
    "rowLevel": Object {
      "allowAll": true,
    },
  },
  Object {
    "conditions": Array [
      Object {
        "if": [Function],
      },
    ],
    "memberLevel": Object {
      "excludes": Array [
        "status",
      ],
      "excludesMembers": Array [
        "ordersExt.status",
      ],
      "includes": "*",
      "includesMembers": Array [
        "ordersExt.count",
        "ordersExt.count_distinct",
        "ordersExt.id",
        "ordersExt.user_id",
        "ordersExt.status",
        "ordersExt.created_at",
        "ordersExt.completed_at",
        "ordersExt.city",
        "ordersExt.sfUsers",
        "ordersExt.anotherStatus",
      ],
    },
    "role": "admin",
    "rowLevel": Object {
      "filters": Array [
        Object {
          "member": [Function],
          "memberReference": "ordersExt.id",
          "operator": "equals",
          "values": [Function],
        },
      ],
    },
  },
  Object {
    "conditions": Array [
      Object {
        "if": [Function],
      },
    ],
    "memberLevel": Object {
      "excludes": Array [
        "min",
        "max",
      ],
      "excludesMembers": Array [
        "ordersExt.min",
        "ordersExt.max",
      ],
      "includes": "*",
      "includesMembers": Array [
        "ordersExt.count",
        "ordersExt.count_distinct",
        "ordersExt.id",
        "ordersExt.user_id",
        "ordersExt.status",
        "ordersExt.created_at",
        "ordersExt.completed_at",
        "ordersExt.city",
        "ordersExt.sfUsers",
        "ordersExt.anotherStatus",
      ],
    },
    "role": "manager",
    "rowLevel": Object {
      "filters": Array [
        Object {
          "or": Array [
            Object {
              "member": [Function],
              "memberReference": "ordersExt.location",
              "operator": "startsWith",
              "values": [Function],
            },
            Object {
              "member": [Function],
              "memberReference": "ordersExt.location",
              "operator": "startsWith",
              "values": [Function],
            },
          ],
        },
      ],
    },
  },
]
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 1`] = `
Object {
  "completed_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "created_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "id": Object {
    "ownedByCube": true,
    "primaryKey": true,
    "sql": [Function],
    "type": "number",
  },
  "status": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "user_id": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "number",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 2`] = `
Object {
  "city": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "completed_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "created_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "id": Object {
    "ownedByCube": true,
    "primaryKey": true,
    "sql": [Function],
    "type": "number",
  },
  "status": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "user_id": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "number",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 3`] = `
Object {
  "count": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "count",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 4`] = `
Object {
  "count": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "count",
  },
  "count_distinct": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "countDistinct",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 5`] = `
Object {
  "sfUsers": Object {
    "description": "SF users segment from createCubeSchema",
    "ownedByCube": true,
    "sql": [Function],
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 6`] = `
Object {
  "anotherStatus": Object {
    "description": "Just another one",
    "ownedByCube": true,
    "sql": [Function],
  },
  "sfUsers": Object {
    "description": "SF users segment from createCubeSchema",
    "ownedByCube": true,
    "sql": [Function],
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 7`] = `
Object {
  "hello": Object {
    "levels": [Function],
    "title": "World",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 8`] = `
Object {
  "ehlo": Object {
    "levels": [Function],
    "title": "UnderGround",
  },
  "hello": Object {
    "levels": [Function],
    "title": "World",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 9`] = `
Object {
  "countCreatedAt": Object {
    "allowNonStrictDateRangeMatch": true,
    "external": true,
    "granularity": "day",
    "measureReferences": [Function],
    "partitionGranularity": "month",
    "refreshKey": Object {
      "every": "1 hour",
    },
    "scheduledRefresh": true,
    "timeDimensionReference": [Function],
    "type": "rollup",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 10`] = `
Object {
  "countCreatedAt": Object {
    "allowNonStrictDateRangeMatch": true,
    "external": true,
    "granularity": "day",
    "measureReferences": [Function],
    "partitionGranularity": "month",
    "refreshKey": Object {
      "every": "1 hour",
    },
    "scheduledRefresh": true,
    "timeDimensionReference": [Function],
    "type": "rollup",
  },
  "mainPreAggs": Object {
    "dimensionReferences": [Function],
    "external": true,
    "measureReferences": [Function],
    "scheduledRefresh": true,
    "type": "rollup",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 11`] = `
Array [
  Object {
    "role": "common",
    "rowLevel": Object {
      "allowAll": true,
    },
  },
  Object {
    "conditions": Array [
      Object {
        "if": [Function],
      },
    ],
    "memberLevel": Object {
      "excludesMembers": Array [],
      "includes": Array [
        "status",
      ],
      "includesMembers": Array [
        "ordersExt.status",
      ],
    },
    "role": "admin",
    "rowLevel": Object {
      "filters": Array [
        Object {
          "member": [Function],
          "memberReference": "ordersExt.status",
          "operator": "equals",
          "values": [Function],
        },
        Object {
          "or": Array [
            Object {
              "member": [Function],
              "memberReference": "ordersExt.created_at",
              "operator": "notInDateRange",
              "values": [Function],
            },
            Object {
              "member": [Function],
              "memberReference": "ordersExt.created_at",
              "operator": "equals",
              "values": [Function],
            },
          ],
        },
        Object {
          "and": Array [
            Object {
              "member": [Function],
              "memberReference": "ordersExt.completed_at",
              "operator": "notInDateRange",
              "values": [Function],
            },
            Object {
              "member": [Function],
              "memberReference": "ordersExt.completed_at",
              "operator": "equals",
              "values": [Function],
            },
          ],
        },
      ],
    },
  },
]
`;

exports[`Schema Testing Inheritance CubeB.js correctly extends cubeA.yml (with additions) 12`] = `
Array [
  Object {
    "role": "common",
    "rowLevel": Object {
      "allowAll": true,
    },
  },
  Object {
    "conditions": Array [
      Object {
        "if": [Function],
      },
    ],
    "memberLevel": Object {
      "excludesMembers": Array [],
      "includes": Array [
        "status",
      ],
      "includesMembers": Array [
        "ordersExt.status",
      ],
    },
    "role": "admin",
    "rowLevel": Object {
      "filters": Array [
        Object {
          "member": [Function],
          "memberReference": "ordersExt.status",
          "operator": "equals",
          "values": [Function],
        },
        Object {
          "or": Array [
            Object {
              "member": [Function],
              "memberReference": "ordersExt.created_at",
              "operator": "notInDateRange",
              "values": [Function],
            },
            Object {
              "member": [Function],
              "memberReference": "ordersExt.created_at",
              "operator": "equals",
              "values": [Function],
            },
          ],
        },
        Object {
          "and": Array [
            Object {
              "member": [Function],
              "memberReference": "ordersExt.completed_at",
              "operator": "notInDateRange",
              "values": [Function],
            },
            Object {
              "member": [Function],
              "memberReference": "ordersExt.completed_at",
              "operator": "equals",
              "values": [Function],
            },
          ],
        },
      ],
    },
  },
  Object {
    "conditions": Array [
      Object {
        "if": [Function],
      },
    ],
    "memberLevel": Object {
      "excludes": Array [
        "min",
        "max",
      ],
      "excludesMembers": Array [
        "ordersExt.min",
        "ordersExt.max",
      ],
      "includes": "*",
      "includesMembers": Array [
        "ordersExt.count",
        "ordersExt.count_distinct",
        "ordersExt.id",
        "ordersExt.user_id",
        "ordersExt.status",
        "ordersExt.created_at",
        "ordersExt.completed_at",
        "ordersExt.city",
        "ordersExt.sfUsers",
        "ordersExt.anotherStatus",
      ],
    },
    "role": "manager",
    "rowLevel": Object {
      "filters": Array [
        Object {
          "or": Array [
            Object {
              "member": [Function],
              "memberReference": "ordersExt.location",
              "operator": "startsWith",
              "values": [Function],
            },
            Object {
              "member": [Function],
              "memberReference": "ordersExt.location",
              "operator": "startsWith",
              "values": [Function],
            },
          ],
        },
      ],
    },
  },
]
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 1`] = `
Object {
  "completed_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "created_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "id": Object {
    "ownedByCube": true,
    "primaryKey": true,
    "sql": [Function],
    "type": "number",
  },
  "status": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "user_id": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "number",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 2`] = `
Object {
  "city": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "completed_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "created_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "id": Object {
    "ownedByCube": true,
    "primaryKey": true,
    "sql": [Function],
    "type": "number",
  },
  "status": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "user_id": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "number",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 3`] = `
Object {
  "count": Object {
    "ownedByCube": true,
    "type": "count",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 4`] = `
Object {
  "count": Object {
    "ownedByCube": true,
    "type": "count",
  },
  "count_distinct": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "countDistinct",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 5`] = `
Object {
  "sfUsers": Object {
    "description": "SF users segment from createCubeSchema",
    "ownedByCube": true,
    "sql": [Function],
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 6`] = `
Object {
  "anotherStatus": Object {
    "description": "Just another one",
    "ownedByCube": true,
    "sql": [Function],
  },
  "sfUsers": Object {
    "description": "SF users segment from createCubeSchema",
    "ownedByCube": true,
    "sql": [Function],
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 7`] = `
Object {
  "hello": Object {
    "levels": [Function],
    "title": "World",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 8`] = `
Object {
  "ehlo": Object {
    "levels": [Function],
    "title": "UnderGround",
  },
  "hello": Object {
    "levels": [Function],
    "title": "World",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 9`] = `
Object {
  "countCreatedAt": Object {
    "external": true,
    "granularity": "day",
    "measureReferences": [Function],
    "partitionGranularity": "month",
    "refreshKey": Object {
      "every": "1 hour",
    },
    "scheduledRefresh": true,
    "timeDimensionReference": [Function],
    "type": "rollup",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 10`] = `
Object {
  "countCreatedAt": Object {
    "external": true,
    "granularity": "day",
    "measureReferences": [Function],
    "partitionGranularity": "month",
    "refreshKey": Object {
      "every": "1 hour",
    },
    "scheduledRefresh": true,
    "timeDimensionReference": [Function],
    "type": "rollup",
  },
  "mainPreAggs": Object {
    "dimensionReferences": [Function],
    "external": true,
    "measureReferences": [Function],
    "scheduledRefresh": true,
    "type": "rollup",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 11`] = `
Array [
  Object {
    "role": "*",
    "rowLevel": Object {
      "allowAll": true,
    },
  },
  Object {
    "conditions": Array [
      Object {
        "if": [Function],
      },
    ],
    "memberLevel": Object {
      "excludes": Array [
        "status",
      ],
      "excludesMembers": Array [
        "ordersExt.status",
      ],
      "includes": "*",
      "includesMembers": Array [
        "ordersExt.count",
        "ordersExt.count_distinct",
        "ordersExt.id",
        "ordersExt.user_id",
        "ordersExt.status",
        "ordersExt.created_at",
        "ordersExt.completed_at",
        "ordersExt.city",
        "ordersExt.sfUsers",
        "ordersExt.anotherStatus",
      ],
    },
    "role": "admin",
    "rowLevel": Object {
      "filters": Array [
        Object {
          "member": [Function],
          "memberReference": "ordersExt.id",
          "operator": "equals",
          "values": [Function],
        },
      ],
    },
  },
]
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.js (with additions) 12`] = `
Array [
  Object {
    "role": "*",
    "rowLevel": Object {
      "allowAll": true,
    },
  },
  Object {
    "conditions": Array [
      Object {
        "if": [Function],
      },
    ],
    "memberLevel": Object {
      "excludes": Array [
        "status",
      ],
      "excludesMembers": Array [
        "ordersExt.status",
      ],
      "includes": "*",
      "includesMembers": Array [
        "ordersExt.count",
        "ordersExt.count_distinct",
        "ordersExt.id",
        "ordersExt.user_id",
        "ordersExt.status",
        "ordersExt.created_at",
        "ordersExt.completed_at",
        "ordersExt.city",
        "ordersExt.sfUsers",
        "ordersExt.anotherStatus",
      ],
    },
    "role": "admin",
    "rowLevel": Object {
      "filters": Array [
        Object {
          "member": [Function],
          "memberReference": "ordersExt.id",
          "operator": "equals",
          "values": [Function],
        },
      ],
    },
  },
  Object {
    "memberLevel": Object {
      "excludes": Array [
        "status",
      ],
      "excludesMembers": Array [
        "ordersExt.status",
      ],
      "includesMembers": Array [
        "ordersExt.count",
        "ordersExt.count_distinct",
        "ordersExt.id",
        "ordersExt.user_id",
        "ordersExt.status",
        "ordersExt.created_at",
        "ordersExt.completed_at",
        "ordersExt.city",
        "ordersExt.sfUsers",
        "ordersExt.anotherStatus",
      ],
    },
    "role": "manager",
  },
]
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 1`] = `
Object {
  "completed_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "created_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "id": Object {
    "ownedByCube": true,
    "primaryKey": true,
    "sql": [Function],
    "type": "number",
  },
  "status": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "user_id": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "number",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 2`] = `
Object {
  "city": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "completed_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "created_at": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "time",
  },
  "id": Object {
    "ownedByCube": true,
    "primaryKey": true,
    "sql": [Function],
    "type": "number",
  },
  "status": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "string",
  },
  "user_id": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "number",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 3`] = `
Object {
  "count": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "count",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 4`] = `
Object {
  "count": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "count",
  },
  "count_distinct": Object {
    "ownedByCube": true,
    "sql": [Function],
    "type": "countDistinct",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 5`] = `
Object {
  "sfUsers": Object {
    "description": "SF users segment from createCubeSchema",
    "ownedByCube": true,
    "sql": [Function],
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 6`] = `
Object {
  "anotherStatus": Object {
    "description": "Just another one",
    "ownedByCube": true,
    "sql": [Function],
  },
  "sfUsers": Object {
    "description": "SF users segment from createCubeSchema",
    "ownedByCube": true,
    "sql": [Function],
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 7`] = `
Object {
  "hello": Object {
    "levels": [Function],
    "title": "World",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 8`] = `
Object {
  "ehlo": Object {
    "levels": [Function],
    "title": "UnderGround",
  },
  "hello": Object {
    "levels": [Function],
    "title": "World",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 9`] = `
Object {
  "countCreatedAt": Object {
    "allowNonStrictDateRangeMatch": true,
    "external": true,
    "granularity": "day",
    "measureReferences": [Function],
    "partitionGranularity": "month",
    "refreshKey": Object {
      "every": "1 hour",
    },
    "scheduledRefresh": true,
    "timeDimensionReference": [Function],
    "type": "rollup",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 10`] = `
Object {
  "countCreatedAt": Object {
    "allowNonStrictDateRangeMatch": true,
    "external": true,
    "granularity": "day",
    "measureReferences": [Function],
    "partitionGranularity": "month",
    "refreshKey": Object {
      "every": "1 hour",
    },
    "scheduledRefresh": true,
    "timeDimensionReference": [Function],
    "type": "rollup",
  },
  "mainPreAggs": Object {
    "dimensionReferences": [Function],
    "external": true,
    "measureReferences": [Function],
    "scheduledRefresh": true,
    "type": "rollup",
  },
}
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 11`] = `
Array [
  Object {
    "role": "common",
    "rowLevel": Object {
      "allowAll": true,
    },
  },
  Object {
    "conditions": Array [
      Object {
        "if": [Function],
      },
    ],
    "memberLevel": Object {
      "excludesMembers": Array [],
      "includes": Array [
        "status",
      ],
      "includesMembers": Array [
        "ordersExt.status",
      ],
    },
    "role": "admin",
    "rowLevel": Object {
      "filters": Array [
        Object {
          "member": [Function],
          "memberReference": "ordersExt.status",
          "operator": "equals",
          "values": [Function],
        },
        Object {
          "or": Array [
            Object {
              "member": [Function],
              "memberReference": "ordersExt.created_at",
              "operator": "notInDateRange",
              "values": [Function],
            },
            Object {
              "member": [Function],
              "memberReference": "ordersExt.created_at",
              "operator": "equals",
              "values": [Function],
            },
          ],
        },
        Object {
          "and": Array [
            Object {
              "member": [Function],
              "memberReference": "ordersExt.completed_at",
              "operator": "notInDateRange",
              "values": [Function],
            },
            Object {
              "member": [Function],
              "memberReference": "ordersExt.completed_at",
              "operator": "equals",
              "values": [Function],
            },
          ],
        },
      ],
    },
  },
]
`;

exports[`Schema Testing Inheritance CubeB.yml correctly extends cubeA.yml (with additions) 12`] = `
Array [
  Object {
    "role": "common",
    "rowLevel": Object {
      "allowAll": true,
    },
  },
  Object {
    "conditions": Array [
      Object {
        "if": [Function],
      },
    ],
    "memberLevel": Object {
      "excludesMembers": Array [],
      "includes": Array [
        "status",
      ],
      "includesMembers": Array [
        "ordersExt.status",
      ],
    },
    "role": "admin",
    "rowLevel": Object {
      "filters": Array [
        Object {
          "member": [Function],
          "memberReference": "ordersExt.status",
          "operator": "equals",
          "values": [Function],
        },
        Object {
          "or": Array [
            Object {
              "member": [Function],
              "memberReference": "ordersExt.created_at",
              "operator": "notInDateRange",
              "values": [Function],
            },
            Object {
              "member": [Function],
              "memberReference": "ordersExt.created_at",
              "operator": "equals",
              "values": [Function],
            },
          ],
        },
        Object {
          "and": Array [
            Object {
              "member": [Function],
              "memberReference": "ordersExt.completed_at",
              "operator": "notInDateRange",
              "values": [Function],
            },
            Object {
              "member": [Function],
              "memberReference": "ordersExt.completed_at",
              "operator": "equals",
              "values": [Function],
            },
          ],
        },
      ],
    },
  },
  Object {
    "memberLevel": Object {
      "excludes": Array [
        "status",
      ],
      "excludesMembers": Array [
        "ordersExt.status",
      ],
      "includesMembers": Array [
        "ordersExt.count",
        "ordersExt.count_distinct",
        "ordersExt.id",
        "ordersExt.user_id",
        "ordersExt.status",
        "ordersExt.created_at",
        "ordersExt.completed_at",
        "ordersExt.city",
        "ordersExt.sfUsers",
        "ordersExt.anotherStatus",
      ],
    },
    "role": "manager",
  },
]
`;
